ProjFS.Mac/PrjFSKextTests/HandleOperationTests.mm

```diff
@@ -0,0 +1,569 @@
+}
+
+- (void) testEventsAreIgnored {
+    testDirVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot | FileFlags_IsEmpty;
+    const int actionCount = 5;
+    kauth_action_t actions[actionCount] =
+    {
+        KAUTH_VNODE_APPEND_DATA,
+        KAUTH_VNODE_ADD_SUBDIRECTORY,
+        KAUTH_VNODE_WRITE_SECURITY,
+        KAUTH_VNODE_TAKE_OWNERSHIP,
+        KAUTH_VNODE_ACCESS
+    };
+    
+    for (int i = 0; i < actionCount; i++)
+    {
+        // Verify for File node
+        HandleVnodeOperation(
+            nullptr,
+            nullptr,
+            actions[i],
+            reinterpret_cast<uintptr_t>(context),
+            reinterpret_cast<uintptr_t>(testFileVnode.get()),
+            0,
+            0);
+        XCTAssertFalse(
+            MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+        MockCalls::Clear();
+
+        // Verify for Directory node
+        HandleVnodeOperation(
+            nullptr,
+            nullptr,
+            actions[i],
+            reinterpret_cast<uintptr_t>(context),
+            reinterpret_cast<uintptr_t>(testDirVnode.get()),
+            0,
+            0);
+        XCTAssertFalse(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+        MockCalls::Clear();
+    }
+}
+
+- (void) testOpen {
+    // A file that is not yet in the virtual root, should recieve a creation request
+    testFileVnode->attrValues.va_flags = 0;
+    HandleFileOpOperation(
+        nullptr,
+        nullptr,
+        KAUTH_FILEOP_OPEN,
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        reinterpret_cast<uintptr_t>(filePath),
+        0,
+        0);
+    
+    XCTAssertTrue(
+       MockCalls::DidCallFunction(
+            ProviderMessaging_TrySendRequestAndWaitForResponse,
+            _,
+            MessageType_KtoU_NotifyFileCreated,
+            testFileVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            _));
+}
+
+- (void) testOpenInVirtualizationRoot {
+    // A file that is already in the virtual root, should not recieve a creation request
+    testFileVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+
+    HandleFileOpOperation(
+        nullptr,
+        nullptr,
+        KAUTH_FILEOP_OPEN,
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        reinterpret_cast<uintptr_t>(filePath),
+        0,
+        0);
+    
+    XCTAssertFalse(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+}
+
+- (void) testCloseWithModifed {
+    testFileVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+
+    HandleFileOpOperation(
+        nullptr,
+        nullptr,
+        KAUTH_FILEOP_CLOSE,
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        reinterpret_cast<uintptr_t>(filePath),
+        KAUTH_FILEOP_CLOSE_MODIFIED,
```

Comment:
I just noticed in the code this is testing that `KAUTH_FILEOP_CLOSE_MODIFIED` is in theory a bit field, but `HandleFileOpOperation()` tests for equality. There currently aren't any other possible flags, but if Apple were to add some in future, our code could break. So (1) we should replace `if (KAUTH_FILEOP_CLOSE_MODIFIED != closeFlags)` with `if ((KAUTH_FILEOP_CLOSE_MODIFIED & closeFlags) == 0)` (or helper function call to that effect) and (2) we should add a test case that deliberately invents an extra bit flag that we or (`|`) with `KAUTH_FILEOP_CLOSE_MODIFIED` and make sure the test still passes. (Which it currently won't, but should after the fix.)

Reply:
Would it make sense to reuse ActionBitIsSet for this check?

I think this is a good suggestion for another PR.  I entered an issue to track https://github.com/Microsoft/VFSForGit/issues/971